### PR TITLE
[1.1]Refine detection mAP in metrics.py.

### PR DESCRIPTION
**Pull request by qingqing01**
evaluator.py throws warnings and tell users to use DetectionMAP in metrics.py, but it is wrong in metrics.py. So refine this API in metrics.py.

This change has been tested in MobileNet-SSD model.

### CODE REVIEW

**Review by dzhwinter — DISMISSED** [review]
Nice Catch.

File: python/paddle/fluid/metrics.py [hunk 1]
```diff
@@ -616,3 +555,185 @@ def eval(self):
...
+        ap_version (string): The average precision calculation ways, it must be
+            'integral' or '11point'. Please check
+            https://sanchom.wordpress.com/tag/average-precision/ for details.
+            - 11point: the 11-point interpolated average precision.
+            - integral: the natural integral of the precision-recall curve.
+
+    Examples:
+        .. code-block:: python
+
+            exe = fluid.executor(place)
```
Comment (panyx0718) [inline python/paddle/fluid/metrics.py]: `Executor`

File: python/paddle/fluid/metrics.py [hunk 2]
```diff
@@ -616,3 +555,185 @@ def eval(self):
...
+        reset metric states at the begin of each pass/user specified batch.
+
+        Args:
+            executor(Executor|ParallelExecutor): a executor for executing
+                the reset_program.
+            reset_program(Program|None): a single Program for reset process.
+                If None, will create a Program.
+        """
+        from .framework import Program, Variable, program_guard
+        from . import layers
```
Comment (panyx0718) [inline python/paddle/fluid/metrics.py]: import at the top?

File: python/paddle/fluid/metrics.py [hunk 3]
```diff
@@ -616,3 +555,185 @@ def eval(self):
...
+                lod_level=var.lod_level,
+                persistable=var.persistable)
+
+        if reset_program is None:
+            reset_program = Program()
+        with program_guard(main_program=reset_program):
+            var = _clone_var_(reset_program.current_block(), self.has_state)
+            layers.fill_constant(
+                shape=var.shape, value=0, dtype=var.dtype, out=var)
+        executor.run(reset_program)
```
Comment (panyx0718) [inline python/paddle/fluid/metrics.py:737]: Doesn't support ParallelExecutor?

File: python/paddle/fluid/metrics.py [hunk 4]
```diff
@@ -616,3 +555,185 @@ def eval(self):
...
+            accumulative mAP variable cross mini-batches.
+        """
+        return self.cur_map, self.accum_map
+
+    def reset(self, executor, reset_program=None):
+        """
+        reset metric states at the begin of each pass/user specified batch.
+
+        Args:
+            executor(Executor|ParallelExecutor): a executor for executing
```
Comment (panyx0718) [inline python/paddle/fluid/metrics.py]: Doesn't support ParallelExecutor?

File: python/paddle/fluid/metrics.py [hunk 5]
```diff
@@ -616,3 +555,185 @@ def eval(self):
...
+                 gt_box,
+                 gt_difficult=None,
+                 class_num=None,
+                 background_label=0,
+                 overlap_threshold=0.5,
+                 evaluate_difficult=True,
+                 ap_version='integral'):
+        from . import layers
+        from .layer_helper import LayerHelper
+        from .initializer import Constant
```
Comment (panyx0718) [inline python/paddle/fluid/metrics.py]: import at the top?

**Review by panyx0718 — COMMENTED** [review]
Is there a test somewhere for MAP?

File: python/paddle/fluid/metrics.py (hunk 1 above)
Comment (qingqing01) [reply python/paddle/fluid/metrics.py]: Done. Thanks!

File: python/paddle/fluid/metrics.py (hunk 5 above)
Comment (qingqing01) [reply python/paddle/fluid/metrics.py]: Move to top.

File: python/paddle/fluid/metrics.py (hunk 2 above)
Comment (qingqing01) [reply python/paddle/fluid/metrics.py]: Done.

File: python/paddle/fluid/metrics.py (hunk 3 above)
Comment (qingqing01) [reply python/paddle/fluid/metrics.py:737]: This evalutor need to merge resluts at first, then calculate mAP. So executor is enough

File: python/paddle/fluid/metrics.py (hunk 4 above)
Comment (qingqing01) [reply python/paddle/fluid/metrics.py]: This evalutor need to merge resluts at first, then calculate mAP. So executor is enough

**Review by qingqing01 — COMMENTED** [review]
@panyx0718 add unit testing.

File: python/paddle/fluid/metrics.py (hunk 4 above)
Comment (panyx0718) [reply python/paddle/fluid/metrics.py]: Right. I mean you should remove `ParallelExecutor` in comment so people won't pass in a ParallelExecutor.

**Review by panyx0718 — DISMISSED** [review]

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**